File: modules/system-integration/android/src/main/java/expo/modules/systemintegration/SystemIntegrationModule.kt
package expo.modules.systemintegration

import android.content.Context
import android.os.Build
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class SystemIntegrationModule : Module() {
  private val context: Context get() = requireNotNull(appContext.reactContext).applicationContext
  private var stopObserving: (() -> Unit)? = null

  override fun definition() = ModuleDefinition {
    Name("SystemIntegration")
    Events("onPending", "onIntentCancelled")
    OnCreate { stopObserving = SystemEntryStore.observe { sendEvent("onPending") } }
    OnDestroy { stopObserving?.invoke(); stopObserving = null }

    Function("getCapabilities") {
      mapOf("translationWindow" to true, "translationProvider" to false, "translationShortcut" to false, "shortcuts" to (Build.VERSION.SDK_INT >= 25))
    }
    AsyncFunction("invalidateTranslationConfiguration") Coroutine { ->
      withContext(Dispatchers.IO) { TranslationConfigurationStore.invalidate(context) }
    }
    AsyncFunction("publishTranslationConfiguration") Coroutine { configuration: Map<String, Any?>, apiKey: String ->
      withContext(Dispatchers.IO) { TranslationConfigurationStore.publish(context, configuration, apiKey) }
    }
    AsyncFunction("publishTranslationUnavailable") Coroutine { configuration: Map<String, Any?> ->
      withContext(Dispatchers.IO) { TranslationConfigurationStore.publishUnavailable(context, configuration) }
    }
    AsyncFunction("getTranslationRevision") Coroutine { ->
      withContext(Dispatchers.IO) { TranslationConfigurationStore.revision(context) }
    }
    AsyncFunction("claimNextEntry") Coroutine { -> withContext(Dispatchers.IO) { SystemEntryStore.claimNext(context) } }
    AsyncFunction("releaseEntry") { id: String -> SystemEntryStore.release(id) }
    AsyncFunction("completeEntry") Coroutine { id: String -> withContext(Dispatchers.IO) { SystemEntryStore.complete(context, id) } }
    AsyncFunction("finishIntent") { _: String, _: Map<String, Any?> -> Unit }
    AsyncFunction("publishAgents") { _: List<Map<String, String>> -> Unit }
  }
}

File: modules/system-integration/ios/SystemIntegrationModule.swift
import ExpoModulesCore
import Foundation

public class SystemIntegrationModule: Module {
  private var observers: [NSObjectProtocol] = []

  public func definition() -> ModuleDefinition {
    Name("SystemIntegration")
    Events("onPending", "onIntentCancelled")
    OnCreate {
      self.observers = [
        NotificationCenter.default.addObserver(forName: SystemEntryStore.pendingNotification, object: nil, queue: .main) { [weak self] _ in self?.sendEvent("onPending") },
        NotificationCenter.default.addObserver(forName: SystemEntryStore.cancelledNotification, object: nil, queue: .main) { [weak self] event in
          if let id = event.userInfo?["id"] as? String { self?.sendEvent("onIntentCancelled", ["id": id]) }
        }
      ]
    }
    OnDestroy { self.observers.forEach { NotificationCenter.default.removeObserver($0) }; self.observers = [] }
    AsyncFunction("claimNextEntry") { try SystemEntryStore.claimNext() }
    AsyncFunction("releaseEntry") { (id: String) in SystemEntryStore.release(id) }
    AsyncFunction("completeEntry") { (id: String) in try SystemEntryStore.complete(id) }
    AsyncFunction("finishIntent") { (id: String, result: [String: Any]) in SystemEntryStore.finishIntent(id, result: result) }
    AsyncFunction("publishAgents") { (agents: [[String: String]]) in
      let metadata = agents.prefix(1000).compactMap { agent -> [String: String]? in
        guard let id = agent["id"], !id.isEmpty, id.count <= 512, let name = agent["name"] else { return nil }
        return ["id": id, "name": String(name.prefix(255))]
      }
      try SystemIntegrationFiles.write(metadata, to: SystemIntegrationFiles.root().appendingPathComponent("agents.json"))
    }
  }
}
